src/main/java/com/google/sps/database/DatabaseImpl.java

```diff
@@ -0,0 +1,202 @@
+
+    }
+
+    //================================================================================
+    // Public Interface Functions (see DatabaseInterface.java for official descriptions)
+    //================================================================================    
+
+    // add a video to the database
+    public int addVideo(String videoID, String metadata) { 
+        Entity vidEnt = new Entity(vKind, videoID); //error possibility
+        datastore.put(vidEnt);  //error possibility
+
+        Entity metaEnt = new Entity(mKind, metadata, vidEnt.getKey()); //error possibility
+        datastore.put(metaEnt); //error possibility
+
+        return 0;
+    }
+
+    // add a keyphrase + timestamp pair to a particular video in the db
+    public int addClause(String videoID, String keyword, List<Long> timestamps) {
+        Entity vidEnt = getVideo(videoID);  //error possibility
+        Entity keyEnt = new Entity(cKind, keyword, vidEnt.getKey()); //error possibility
+        keyEnt.setProperty(tKind, timestamps); //error possibility
+        datastore.put(keyEnt); //error possibility
+
+        return 0;
+    }
+    
+    // add multiple keyphrase + timestamp pairs to a particular video's entry
+    public int addClauses(String videoID, Map<String, List<Long>> clauses) {
+        for (Map.Entry<String, List<Long>> me : clauses.entrySet()) {
+            addClause(videoID, me.getKey(), me.getValue());  //error possibility
+        }
+
+        return 0;
+    }
+    
+    // add metadata information to a particular video in the db
+    public int addMetadata(String videoID, String metadata, boolean overwrite) {
+        Entity vidEnt = getVideo(videoID); //error possibility
+        Entity metaEnt;
+        
+        // need to find metadata already in the database
+        Entity currMeta = getMetadata(videoID);  //error possibility
+        Key metaKey = currMeta.getKey();  //error possibility
+
+        if (overwrite) {
+            datastore.delete(metaKey); //error possibility
+            metaEnt = new Entity(mKind, metadata, vidEnt.getKey()); //error possibility
+            datastore.put(metaEnt); //error possibility
+        } else {
+// TO BE IMPLEMENTED (do not delete)
+            /*
+            try {
+                metaEnt = datastore.get(metaKey); //error possibility
+            } catch (Exception e) {
+                return -1;
+            }
+            String current = metaEnt.getProperty("ID/Name").toString(); //error possibility
+            datastore.delete(metaKey); //error possibility
+            metaEnt = new Entity(m, current + metadata, vidEnt.getKey()); //error possibility
+            */
+            
+        }
+        //datastore.put(metaEnt); //error possibility
+
+        return 0;
+    }
+    
+    // retrieve all keywords + their timestamps in a specified videoID
+    public Map<String, List<Long>> getAllKeywords(String videoID) {
+        Entity vidEnt = getVideo(videoID);  //error possibility
```

Comment:
If known errors could be thrown, shouldn't your function `throw` these errors?

Reply:
addressing both of your comments: this version does not address errors yet. both the int return type functions & try-catches will be used to address possible errors and all that. I am currently implementing all that and that will be the next database pr after this one gets confirmed.